### ISSUE TITLE
Roadmap to Webpack Encore

### ISSUE THREAD

**Issue by jordisala1991** [sonata-project/SonataAdminBundle#7049]
## Feature Request

We recently moved to NPM to handle all the dependencies of our code: https://github.com/sonata-project/SonataAdminBundle/pull/7034

## First Round

- [x] Integrate / Remove missing dependencies (#7084)
  - [x] Jquery UI (#7084)
  - [x] Eonasdan dateTime picker (#7084)
  - [x] jQuery.confirmExit (#7084)
- [x] Upgrade some dependencies (https://github.com/sonata-project/SonataAdminBundle/pull/7067)
  - [x] Do not fix versions on package.json (https://github.com/sonata-project/SonataAdminBundle/pull/7067)
  - [x] Upgrade Select2 (https://github.com/sonata-project/SonataAdminBundle/pull/7048)
  - [x] Update rest of dependencies (https://github.com/sonata-project/SonataAdminBundle/pull/7067)
  - [x] Replace `components-font-awesome` for `@fortawesome/fontawesome-free` (#7133)
- [x] Remove ionicons (https://github.com/sonata-project/SonataAdminBundle/pull/7067)
- [x] Integrate our code with the webpack encore process (#7109)
  - [x] Move javascript to root assets directory (#7098)
  - [x] Move css to root assets directory (https://github.com/sonata-project/SonataAdminBundle/pull/7073)
  - [x] Move img to root assets directory (#7109)
  - [x] Convert css to scss (https://github.com/sonata-project/SonataAdminBundle/pull/7073)
- [x] Integrate our js files to webpack encore (#7106)
- [x] Minify all css files together (#7084)
- [x] Minify all js files together (#7098)
- [x] Add stylelint (#7081)
  - [x] Add more rules to stylelint ([stylelint-scss](https://www.npmjs.com/package/stylelint-scss), [stylelint-order](https://www.npmjs.com/package/stylelint-order)) (#7113)
  - [x] Lint actual files (#7081)
  - [x] Configure Github actions to lint files when assets directory changes (#7081)
- [x] Add Autoprefixer (#7089)
- [x] Add eslint (#7098)
  - [x] Add more rules to eslint ([eslint-plugin-notice](https://www.npmjs.com/package/eslint-plugin-notice) or [eslint-plugin-header](https://www.npmjs.com/package/eslint-plugin-header)) (#7113)
  - [x] Lint actual files (#7098)
  - [x] Lint with airbnb-base (#7113)
  - [x] Configure Github actions to lint files when assets directory changes (#7081)
- [x] Integrate Source Sans Pro Font (https://github.com/sonata-project/SonataAdminBundle/pull/7131)
- [x] Improve dev-kit to handle projects with javascript and css (https://github.com/sonata-project/dev-kit/pull/1505)
- [x] Documentation about the frontend assets for contributors (https://github.com/sonata-project/dev-kit/pull/1506)
- [ ] Remove inline javascript from templates **(ordered by complexity)**
  - [ ] [base-history](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/base_history.html.twig#L60-L90)
  - [ ] [base-list](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/base_list.html.twig#L142-L171)
  - [ ] [edit-one-script](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/Association/edit_one_script.html.twig#L23-L94)
  - [ ] [edit-one-to-many-sortable-script-table](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/Association/edit_one_to_many_sortable_script_table.html.twig#L11-L40)
  - [ ] [edit-one-to-many-sortable-tabs](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/Association/edit_one_to_many_sortable_script_tabs.html.twig#L11-L44)
  - [ ] [form-admin-fields-1](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/Form/form_admin_fields.html.twig#L480-L525)
  - [ ] [form-admin-fields-2](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/Form/form_admin_fields.html.twig#L531-L539)
  - [ ] [standard-layout](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/standard_layout.html.twig#L333-L335) This one probably should be deprecated
  - [ ] [sonata-type-model-autocomplete](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/Form/Type/sonata_type_model_autocomplete.html.twig#L52-L278) This one will be hard
  - [ ] [edit-many-script](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Resources/views/CRUD/Association/edit_many_script.html.twig#L25-L547) This one will be hard
- [ ] Documentation about the frontend assets for users of the bundle

## Second round

Major upgrades

- [ ] Replace `eonasdan-bootstrap-datetimepicker` for `tempusdominus-bootstrap-4`
- [ ] Update jQuery to v3

Upgrade to Admin-lte v3

- [ ] Remove x-editable (or make it compatible with boostrap 4 or implement own solution)
- [ ] Upgrade bootstrap to v4
- [ ] Upgrade admin lte to v3 (check sandbox branch and older prs on sonata admin)

Not needed to close this issue

- [ ] Consider integration with webpack encore bundle

This one is focused on the quality of our code

- [ ] Split css files into smaller components
- [ ] Reduce specifity of css selectors
- [ ] Split js into smaller components
- [ ] Remove usages of jQuery as much as possible

This one will come later, once the first round is done.

**Comment by VincentLanglet** [comment]
It's not clear to me:
If this big to-do-list couldn't be finished (quickly/in time), what is required for a new alpha or the stable version ?

**Comment by jordisala1991** [comment]
Would be nice for the stable version, but not mandatory. To be noted that some of this will break bc, so if we do it after the stable, will need to be done for 5.0.

For example:

Is it mandatory to have admin lte v3 to release sonata admin v4 stable? Not to me, it would be nice but not mandatory.
If we don't get there in time, it will have to wait for 5.0 (or done in a bc way)

**Comment by VincentLanglet** [comment]
> Is it mandatory to have admin lte v3 to release sonata admin v4 stable? Not to me, it would be nice but not mandatory.

If I remember correctly, Admin-lte v3 seems to be a huge change because it will use bootstrap 4 instead of 3.
There is already a lot of change for the v4. So IMHO it's better to do it for the v5.

I'm ok to release the stable v4 in 1-2 month and to release the v5 at the end of the year.
We should release a new major every year after the v4 to avoid this situation again.

**Comment by jordisala1991** [comment]
Im okay with not upgrading admin-lte for v4.

This issue is about making changes that bring value and not blocking releases or try to upgrade a lot of things in a single PR.

**Comment by jordisala1991** [comment]
We have a problem with this dependency: Eonasdan dateTime picker

It is installed here (previously it was on core-bundle), but the date form types are actually in https://github.com/sonata-project/form-extensions/tree/1.x/src/Type

We have 2 solutions here:

* Keep Eonasdan here but with the problem that I cannot update it because it requires to change options in: https://github.com/sonata-project/form-extensions/blob/1.x/src/Type/BasePickerType.php (I don't think it is a good idea)
* Move Eonasdan to the form-extension bridge and modify the resource route (we will still have the update problem, but we can upgrade it for the 2.0 of form-extension)

The only thing with the second option is that we need to configure webpack encore there too.

wdyt? @sonata-project/contributors

**Comment by haivala** [comment]
I have installed bootsratp 4 for the front of my project with webpack and I'm using "tempusdominus-bootstrap-4": "^5.1.2" with
```
            ->add('deliveryDate', DatePickerType::class, [
                'required' => false,
                'widget' => 'single_text',
                'html5' => true,
                'format' => 'd.M.y',
                'attr' => [
                    'data-toggle'=>"datetimepicker",
                    'data-target'=>"#castingbundle_actorrequest_deliveryDate"
                ],
                'datepicker_use_button' => false,
                'dp_days_of_week_disabled' => [0,6],
                'dp_min_date'              => $date->format('Y-m-d\TH:i:s'),
                'dp_use_current'             => false,
                'dp_show_today' => true,
                'dp_max_date' => $maxdate->format('Y-m-d\TH:i:s'),
                'dp_pick_date' => true

            ])
```
and I'm rendering it with `{{ form_row(form.deliveryDate) }}`
And it works.

I don't know if this is helpful for your problem but I'm just pointing out that it works with the picker type

**Comment by jordisala1991** [comment]
> I have installed bootsratp 4 for the front of my project with webpack and I'm using "tempusdominus-bootstrap-4": "^5.1.2" with
> 
> ```
>             ->add('deliveryDate', DatePickerType::class, [
>                 'required' => false,
>                 'widget' => 'single_text',
>                 'html5' => true,
>                 'format' => 'd.M.y',
>                 'attr' => [
>                     'data-toggle'=>"datetimepicker",
>                     'data-target'=>"#castingbundle_actorrequest_deliveryDate"
>                 ],
>                 'datepicker_use_button' => false,
>                 'dp_days_of_week_disabled' => [0,6],
>                 'dp_min_date'              => $date->format('Y-m-d\TH:i:s'),
>                 'dp_use_current'             => false,
>                 'dp_show_today' => true,
>                 'dp_max_date' => $maxdate->format('Y-m-d\TH:i:s'),
>                 'dp_pick_date' => true
> 
>             ])
> ```
> 
> and I'm rendering it with `{{ form_row(form.deliveryDate) }}`
> And it works.
> 
> I don't know if this is helpful for your problem but I'm just pointing out that it works with the picker type

What version of `eonasdan-bootstrap-datetimepicker` do you use?

**Comment by haivala** [comment]
> What version of `eonasdan-bootstrap-datetimepicker` do you use?

none

**Comment by jordisala1991** [comment]
Oh you replaced `eonasdan-bootstrap-datetimepicker` with `tempusdominus-bootstrap-4`.

Still I don't know if all options of our base date time types are compatible with that package. (Also to be noted that we are still using bootstrap 3 here because of admin-lte v2).

The problem that I am trying to point out is that we are installing a package that gives a features used in another library and that's a problem when trying to upgrade things.

**Comment by franmomu** [comment]
can we maybe remove those options from `form-extensions` and add them from here with [a Form Type Extension](https://symfony.com/doc/4.4/form/create_form_type_extension.html#defining-the-form-type-extension)? like `EonasdanDatePickerTypeExtension` or something like that.

**Comment by jordisala1991** [comment]
> can we maybe remove those options from `form-extensions` and add them from here with [a Form Type Extension](https://symfony.com/doc/4.4/form/create_form_type_extension.html#defining-the-form-type-extension)? like `EonasdanDatePickerTypeExtension` or something like that.

That could be a good thing to try, but not sure about this: https://github.com/sonata-project/form-extensions/blob/1.x/src/Bridge/Symfony/Resources/views/Form/datepicker.html.twig

If those form-extensions types are suposed to work standalone, I am missing something, because there is a coupling with some javascript but the bridge provides none.

**Comment by franmomu** [comment]
> If those form-extensions types are suposed to work standalone, I am missing something, because there is a coupling with some javascript but the bridge provides none.

oops, I thought the templates were here, in that case I guess we should also override them (and not sure how would look like in `form-extensions`, maybe they won't exist).

**Comment by jordisala1991** [comment]
If we remove those templates from the extension, how does it work when used standalone? Not sure about that. 

@sonata-project/contributors wdyt?

**Comment by jordisala1991** [comment]
A lot of the work required is already done or on PRs already opened. I just moved some tasks to second priority

**Comment by VincentLanglet** [comment]
> A lot of the work required is already done or on PRs already opened. I just moved some tasks to second priority

We might take all the first round in 4.0 and set 5.0 milestone for the second round :)

**Comment by wbloszyk** [comment]
> > A lot of the work required is already done or on PRs already opened. I just moved some tasks to second priority
> 
> We might take all the first round in 4.0 and set 5.0 milestone for the second round :)

IMO we should consider integration with webpack encore bundle in 4.0. Then other things from second round can be done in 4.x as experimental another encore build.

**Comment by jordisala1991** [comment]
I have listed what <script> tags we have in our templates, it will be a lot of work. Maybe not all will be complete for 4.0

**Comment by haivala** [comment]
For collection is use `@a2lix/symfony-collection` in my frontend. Don't know if it'll work for admin.

**Comment by jordisala1991** [comment]
> For collection is use `@a2lix/symfony-collection` in my frontend. Don't know if it'll work for admin.

Looks nice to look at how they do it, but IMHO we should not ingrate dependencies that does not look well maintained, with activity, and so on... we could end up with the x-editable drama again.

**Comment by haivala** [comment]
> > For collection is use `@a2lix/symfony-collection` in my frontend. Don't know if it'll work for admin.
> 
> Looks nice to look at how they do it, but IMHO we should not ingrate dependencies that does not look well maintained, with not much activity, not stable release and so on... we could end up with the x-editable drama again.

Sure! Just pointing out not an actual suggestion.

**Comment by jordisala1991** [comment]
I didn't found any good plugin to check for license header in stylelint, do you know one?

It should work in a similar way to https://www.npmjs.com/package/eslint-plugin-header but for stylelint. I saw a few, but not very well maintained or not working.

**Comment by VincentLanglet** [comment]
@jordisala1991 I just discovered that the recent update of admin-lte in 3.x lost the import of the `Source sans pro` font.
I assume it's the same for 4.x.

See 
https://github.com/sonata-project/SonataAdminBundle/pull/7032/files?file-filters%5B%5D=.css&file-filters%5B%5D=.jpg&file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.lock&file-filters%5B%5D=.map#
and
https://github.com/ColorlibHQ/AdminLTE/commit/235481d1d6324c1d4257f86c621700aa7f4a0b7f#diff-8f99965cebd561a4920ebfac6065526717f533d00a8ff79b8221c60bb6f549ff

What is the best way for us to add this in 3.x ? And in 4.x ?

**Comment by jordisala1991** [comment]
I think adding it at the end of <head> tag should work:

`<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,600,700,300italic,400italic,600italic">`

**Comment by VincentLanglet** [comment]
Not sure we have to add something to the doc, if I add it to the `standard_layout`.
But since there is 
```
       {% block stylesheets %}
            {% for stylesheet in sonata_config.getOption('stylesheets', []) %}
                <link rel="stylesheet" href="{{ asset(stylesheet) }}">
            {% endfor %}
        {% endblock %}
```
I was wondering if it should be added to the `standard_layout` template or the option `stylesheets` ; wdyt ?

**Comment by jordisala1991** [comment]
Not sure how to handle it, I was thinking on deprecating the default css and javascript configuration now that we have only 1 file each. Maybe it can be on the template with a block to be overriden?

**Comment by core23** [comment]
> I think adding it at the end of `<head>` tag should work:
> 
> `<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,600,700,300italic,400italic,600italic">`
> 
> If it works, maybe you can add it to our docs?

:-1: 

We should not load ANY third party dependency from an other host. We should load this font via `npm` or use copy & paste

**Comment by haivala** [comment]
> Not sure how to handle it, I was thinking on deprecating the default css and javascript configuration now that we have only 1 file each. Maybe it can be on the template with a block to be overriden?

I think there should be way to add simple css or js hacks to admin via config

**Comment by VincentLanglet** [comment]
> > I think adding it at the end of `<head>` tag should work:
> > `<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,600,700,300italic,400italic,600italic">`
> > If it works, maybe you can add it to our docs?
> 
> 👎
> 
> We should not load ANY third party dependency from an other host. We should load this font via `npm` or use copy & paste

This was the way it was loaded before in the css 
```
@import url(https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,600,700,300italic,400italic,600italic);
```
So adding 
```
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,600,700,300italic,400italic,600italic">
```
is just loading the font the same way in order to fix the BC-break introduce in 3.x with the admin-lte bump.

In 4.x maybe we could do it here: https://github.com/sonata-project/SonataAdminBundle/blob/master/assets/scss/app.scss ?

> > Not sure how to handle it, I was thinking on deprecating the default css and javascript configuration now that we have only 1 file each. Maybe it can be on the template with a block to be overriden?
> 
> I think there should be way to add simple css or js hacks to admin via config

@jordisala1991 wanted to deprecate the default config.
https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/DependencyInjection/Configuration.php#L434
Not the `extra_stylesheets` or `extra_javascripts` one.

**Comment by core23** [comment]
> is just loading the font the same way in order to fix the BC-break introduce in 3.x with the admin-lte bump.

For me, this was a security fix and no BR break. I don't want to use any stuff from google and respect the user privacy.

> Or can we download the font and put them here: https://github.com/sonata-project/SonataAdminBundle/tree/master/src/Resources/public/fonts ?

This is the way to go :+1:

**Comment by VincentLanglet** [comment]
> For me, this was a security fix and no BR break. I don't want to use any stuff from google and respect the user privacy.

There was no information in the changelog about the removal, and the design is suffering from this removal.
Plus, admin-lte didn't remove the font but move them to the `html` instead.
In sonata it was removed from the css but not added to the html. Definitely a BC-break.

> > Or can we download the font and put them here: https://github.com/sonata-project/SonataAdminBundle/tree/master/src/Resources/public/fonts ?

I'm not familiar with this. 
@jordisala1991 @core23 do you have time to make the PR with the fix ?

**Comment by jordisala1991** [comment]
IMO 2 options:

1. Add to the assets, just the static fonts (for 3.x we have to do it this way for sure)
2. On 4.x we can use this: https://github.com/fontsource/fontsource

**Comment by VincentLanglet** [comment]
> IMO 2 options:
> 
> 1. Add to the assets, just the static fonts (for 3.x we have to do it this way for sure)

Can you do it ? :)

**Comment by jordisala1991** [comment]
I just did for master: https://github.com/sonata-project/SonataAdminBundle/pull/7131

Once that is approved, the fonts files can be backported to 3.x and the css necessary too. (on the merge back to master will be removed).

**Comment by jordisala1991** [comment]
I will split the remaining list into smaller issues. The integration with webpack encore is done, now it is time to improve what we have.

**Comment by jordisala1991** [comment]
There are some issues that keep the work on frontend going:

#7158 
#7157 
#7156
